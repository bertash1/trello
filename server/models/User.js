const { Schema, model } = require("mongoose");

const UserSchema = new Schema({
  email: {
    type: String,
    unique: true,
    required: true,
  },
  password: {
    type: String,
    required: true,
  },
  isActivated: {
    type: Boolean,
    default: false,
  },
  activationLink: {
    type: String,
  },
  cards: [{
    type: Schema.Types.ObjectId, 
    ref: "Card"
  }],
})

module.exports = User = model("User", UserSchema);