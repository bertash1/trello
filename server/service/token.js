const jwt = require("jsonwebtoken")
const Token = require("../models/Token")

const generateTokens = (payload) => {
  const accessToken = jwt.sign(payload, process.env.JWT_ACCESS_SECRET, {expiresIn: "30m"});
  const refreshToken = jwt.sign(payload, process.env.JWT_REFRESH_SECRET, {expiresIn: "30d"});

  return {
    accessToken,
    refreshToken,
  }
}

const saveToken = async (userId, refreshToken) => {
  const tokenData = await Token.findOne({user: userId});
  if(tokenData) {
    tokenData.refreshToken = refreshToken;
    return tokenData.save();
  }
  const token = await Token.create({user: userId, refreshToken});
  return token;
}

const removeToken = async (refreshToken) => {
  const tokenData = await Token.deleteOne({refreshToken});
  return tokenData;
}

const findToken = async (refreshToken) => {
  const tokenData = await Token.findOne({refreshToken});
  return tokenData;
}

const validateAccessToken = (token) => {
  try {
    const userData = jwt.verify(token, process.env.JWT_ACCESS_SECRET);
    return userData;
  } catch (error) {
    return null;
  }
}

const getUserData = (token) => {
  try {
    const userData = jwt.verify(token, process.env.JWT_ACCESS_SECRET);
    return userData;
  } catch (error) {
    return null
  }
}

const validateRefreshToken = (token) => {
  try {
    const userData = jwt.verify(token, process.env.JWT_REFRESH_SECRET);
    return userData;
  } catch (error) {
    return null;
  }
}


exports.generateTokens = generateTokens;
exports.saveToken = saveToken;
exports.removeToken = removeToken;
exports.validateAccessToken = validateAccessToken;
exports.validateRefreshToken = validateRefreshToken;
exports.findToken = findToken;
exports.getUserData = getUserData;