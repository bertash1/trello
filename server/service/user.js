const User = require("../models/User");
const bcrypt = require("bcrypt");
const uuid = require("uuid");
const {sendActivationMail} = require("./mail");
const {generateTokens, saveToken, removeToken, findToken, validateRefreshToken} = require("./token");
const ApiError = require('../exceptions/api-error');

const userRegistration = async (email, password) => {
  const candidate = await User.findOne({email});
  if (candidate) {
    throw ApiError.BadRequest("User already exists")
  }

  const hashPassword = await bcrypt.hash(password, 3);
  const activationLink = uuid.v4();

  const user = await User.create({
    email, 
    password: hashPassword, 
    activationLink
  });

  await sendActivationMail(email, `${process.env.API_URL}/api/activate/${activationLink}`);
  const tokens = generateTokens({_id: user._id, email: user.email, isActivated: user.isActivated});
  await saveToken(user._id, tokens.refreshToken)

  return {
    ...tokens,
    user: {
      _id: user._id, 
      email: user.email, 
      isActivated: user.isActivated
    }
  }
}

const activateUser = async (activationLink) => {
  const user = await User.findOne({activationLink});

  if (!user) {
    throw ApiError.BadRequest("Incorrect activation link")
  }
  user.isActivated = true;
  await user.save();
}

const loginUser = async (email, password) => {
  const user = await User.findOne({email});
  if(!user) {
    throw ApiError.BadRequest("User was not found")
  }

  const isPassEquals = await bcrypt.compare(password, user.password);
  if(!isPassEquals) {
    throw ApiError.BadRequest("Incorrect password");
  }

  const tokens = generateTokens({_id: user._id, email: user.email, isActivated: user.isActivated});
  await saveToken(user._id, tokens.refreshToken)

  return {
    ...tokens,
    user: {
      _id: user._id, 
      email: user.email, 
      isActivated: user.isActivated
    }
  }
}

const logoutUser = async (refreshToken) => {
  const token = await removeToken(refreshToken);
  return token;
}

const refreshUserToken = async (refreshToken) => {
  if(!refreshToken) {
    throw ApiError.UnauthorizedError();
  }
  const userData = validateRefreshToken(refreshToken);
  const tokenFromDb = await findToken(refreshToken);
  if(!userData || !tokenFromDb) {
    throw ApiError.UnauthorizedError();
  }

  const user = await User.findById(userData.id)
  const tokens = generateTokens({_id: user._id, email: user.email, isActivated: user.isActivated});
  await saveToken(user._id, tokens.refreshToken)
}


exports.userRegistration = userRegistration;
exports.activateUser = activateUser;
exports.loginUser = loginUser;
exports.logoutUser = logoutUser;
exports.refreshUserToken = refreshUserToken;