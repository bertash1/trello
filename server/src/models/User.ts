import { Schema, model } from "mongoose";
import { IUser } from "../types/types"

const UserSchema = model<IUser>("User", new Schema({
  email: {
    type: String,
    unique: true,
    required: true,
  },
  password: {
    type: String,
    required: true,
  },
  isActivated: {
    type: Boolean,
    default: false,
  },
  activationLink: {
    type: String,
  },
}))

export default UserSchema;