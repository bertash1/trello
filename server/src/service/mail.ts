export {};

const nodemailer = require("nodemailer")

const sendActivationMail = async(to:string, link:string) => {
  const transporter = nodemailer.createTransport({
    host: process.env.SMTP_HOST,
    port: process.env.SMTP_PORT,
    secure: false,
    auth: {
      user: process.env.SMTP_USER,
      pass: process.env.SMTP_PASSWORD,
    }
  })

  await transporter.sendMail({
    from: process.env.SMTP_USER,
    to,
    subject: "Account activation on " + process.env.API_URL,
    text: "",
    html:
    `
    <div>
      <h1>To activate your account click the following link</h1>
      <a href="${link}">${link}</a>
    </div>
    `
  })
}

exports.sendActivationMail = sendActivationMail;