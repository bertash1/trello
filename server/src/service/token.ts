import jwt from "jsonwebtoken"
import Token from "../models/Token"
import { IUserData } from "../types/types"

export const generateTokens = (payload: IUserData) => {
  const accessToken = jwt.sign(
    payload,
    process.env.JWT_ACCESS_SECRET as string,
    { expiresIn: "30m" }
  )
  const refreshToken = jwt.sign(
    payload,
    process.env.JWT_REFRESH_SECRET as string,
    { expiresIn: "30d" }
  )

  return {
    accessToken,
    refreshToken,
  }
}

export const saveToken = async (userId: string, refreshToken: string) => {
  const tokenData = await Token.findOne({ user: userId })
  if (tokenData) {
    tokenData.refreshToken = refreshToken
    return tokenData.save()
  }
  const token = await Token.create({ user: userId, refreshToken })
  console.log(token)
  return token
}

export const removeToken = async (refreshToken: string) => {
  const tokenData = await Token.deleteOne({ refreshToken })
  return tokenData
}

export const findToken = async (refreshToken: string) => {
  const tokenData = await Token.findOne({ refreshToken })
  return tokenData
}

export const validateAccessToken = (token: string) => {
  try {
    const userData = jwt.verify(token, process.env.JWT_ACCESS_SECRET as string)
    return userData
  } catch (error) {
    return null
  }
}

export const getUserData = (token: string) => {
  try {
    const userData = jwt.verify(token, process.env.JWT_ACCESS_SECRET as string)
    return userData
  } catch (error) {
    return null
  }
}

export const validateRefreshToken = (token: string) => {
  try {
    const userData = jwt.verify(token, process.env.JWT_REFRESH_SECRET as string)
    return userData
  } catch (error) {
    return null
  }
}
