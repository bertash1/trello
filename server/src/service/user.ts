import User from "../models/User"
import bcrypt from "bcrypt"
import uuid from "uuid"
import sendActivationMail from "./mail"
import {
  generateTokens,
  saveToken,
  removeToken,
  findToken,
  validateRefreshToken,
} from "./token"
import ApiError from "../exceptions/api-error"
import { JwtPayload } from "jsonwebtoken"
import { IToken } from "../types/types"

export const userRegistration = async (email: string, password: string) => {
  const candidate = await User.findOne({ email })
  if (candidate) {
    throw ApiError.BadRequest("User already exists")
  }

  const hashPassword = await bcrypt.hash(password, 3)
  const activationLink: string = uuid.v4()

  const user = await User.create({
    email,
    password: hashPassword,
    activationLink,
  })

  await sendActivationMail(
    email,
    `${process.env.API_URL}/api/activate/${activationLink}`
  )
  const tokens = generateTokens({
    _id: user._id,
    email: user.email,
    isActivated: user.isActivated,
  })
  await saveToken(user._id, tokens.refreshToken)

  return {
    ...tokens,
    user: {
      _id: user._id,
      email: user.email,
      isActivated: user.isActivated,
    },
  }
}

export const activateUser = async (activationLink: string) => {
  const user = await User.findOne({ activationLink })

  if (!user) {
    throw ApiError.BadRequest("Incorrect activation link")
  }
  user.isActivated = true
  await user.save()
}

export const loginUser = async (email: string, password: string) => {
  const user = await User.findOne({ email })
  if (!user) {
    throw ApiError.BadRequest("User was not found")
  }

  const isPassEquals = await bcrypt.compare(password, user.password)
  if (!isPassEquals) {
    throw ApiError.BadRequest("Incorrect password")
  }

  const tokens = generateTokens({
    _id: user._id,
    email: user.email,
    isActivated: user.isActivated,
  })
  await saveToken(user._id, tokens.refreshToken)

  return {
    ...tokens,
    user: {
      _id: user._id,
      email: user.email,
      isActivated: user.isActivated,
    },
  }
}

export const logoutUser = async (refreshToken: string) => {
  const token = await removeToken(refreshToken)
  return token
}

export const refreshUserToken = async (
  refreshToken: string
): Promise<IToken | null> => {
  if (!refreshToken) {
    throw ApiError.UnauthorizedError()
  }
  const userData = validateRefreshToken(refreshToken) as JwtPayload
  const tokenFromDb = await findToken(refreshToken)
  if (!userData || !tokenFromDb) {
    throw ApiError.UnauthorizedError()
  }

  const user = await User.findById(userData.id)

  if (user) {
    const tokens = generateTokens({
      _id: user._id,
      email: user.email,
      isActivated: user.isActivated,
    })
    const tokenData = await saveToken(user._id, tokens.refreshToken)
    return tokenData
  }
  return null
}
