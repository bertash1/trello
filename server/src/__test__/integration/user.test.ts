/* eslint-disable @typescript-eslint/ban-ts-comment */
import supertest from "supertest"

import createServer from "../../utils/server"
import * as SendActivationMail from "../../service/mail"
import User from "../../models/User"

const app = createServer()

describe("User", () => {
  const userPayload = {
    email: "some@test.com",
    password: "12345",
  }

  const userPayloadWithoutPassword = {
    email: "test1@mail.com",
  }

  const userData = {
    email: "some@test.com",
    isActivated: false,
  }

  const fakeUserResponse = {
    accessToken: expect.any(Object),
    refreshToken: expect.any(Object),
    user: {
      _id: expect.any(Object),
      email: userData.email,
      isActivated: userData.isActivated,
    },
  }

  describe("User registration", () => {
    test("should return a 200 status and user data, handle existing users and return a 400 status", async () => {
      jest
        .spyOn(SendActivationMail, "default")
        //@ts-ignore
        .mockReturnValueOnce(fakeUserResponse)

      const response = await supertest(app)
        .post("/api/registration")
        .send(userPayload)

      expect(response.status).toBe(200)
      expect(response.body).toMatchSnapshot()

      const failedResponse = await supertest(app)
        .post("/api/registration")
        .send(userPayload)

      expect(failedResponse.status).toBe(400)
    })

    test("should handle missing required params and return a 400 status", async () => {
      const response = await supertest(app)
        .post("/api/registration")
        .send({ userPayloadWithoutPassword })

      expect(response.status).toBe(400)
    })

    test("should handle not valid email and return a 400 status", async () => {
      const response = await supertest(app)
        .post("/api/registration")
        .send({ email: "notvalidemail", password: "12345" })

      expect(response.status).toBe(400)
    })

    test("should handle not valid password and return a 400 status", async () => {
      const response = await supertest(app)
        .post("/api/registration")
        .send({ email: "validemail@gmail.com", password: "1" })

      expect(response.status).toBe(400)
    })
  })
  describe("User login", () => {
    test("should return a 200 status and user data", async () => {
      const response = await supertest(app)
        .post("/api/login")
        .send({ email: "email@test.com", password: "12345" })

      expect(response.status).toBe(200)
      expect(response.body).toBeTruthy()
    })

    test("should handle incorrect password and return a 404 status", async () => {
      const response = await supertest(app)
        .post("/api/login")
        .send({ email: "email@test.com", password: "Incorrect password" })

      expect(response.status).toBe(404)
    })

    test("should handle incorrect email and return a 404 status", async () => {
      const response = await supertest(app)
        .post("/api/login")
        .send({ email: "incorrect email", password: "12345" })

      expect(response.status).toBe(404)
    })
  })

  describe("User activation", () => {
    test("should activate user", async () => {
      const user = await User.findOne()
      //@ts-ignore
      expect(user.isActivated).toBe(false)
      await supertest(app)
        //@ts-ignore
        .get(`/api/activate/${user.activationLink}`)
      const activatedUser = await User.findOne()
      //@ts-ignore
      expect(activatedUser?.isActivated).toBe(true)
    })

    test("should handle incorrect activation link and return a 400 status", async () => {
      const { status } = await supertest(app).get("/api/activate/incorrectLink")
      expect(status).toBe(400)
    })
  })
})
