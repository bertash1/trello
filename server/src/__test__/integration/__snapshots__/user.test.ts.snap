// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User User registration should return a 200 status and user data, handle existing users and return a 400 status 1`] = `
Object {
  "accessToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2MWI3NWE4MjBkZGQ2ZTAwYWZmNDE4YjEiLCJlbWFpbCI6InNvbWVAdGVzdC5jb20iLCJpc0FjdGl2YXRlZCI6ZmFsc2UsImlhdCI6MTYzOTQwNjIxMCwiZXhwIjoxNjM5NDA4MDEwfQ.u4sshgjXSpu7kB7j__ZnTchbNqmZrXoFrAN-Z4LShfc",
  "refreshToken": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2MWI3NWE4MjBkZGQ2ZTAwYWZmNDE4YjEiLCJlbWFpbCI6InNvbWVAdGVzdC5jb20iLCJpc0FjdGl2YXRlZCI6ZmFsc2UsImlhdCI6MTYzOTQwNjIxMCwiZXhwIjoxNjQxOTk4MjEwfQ.6XYBY2Enxlm7OuwpI5CiDxA1Gd3WC3kB8K5c4yckCzc",
  "user": Object {
    "_id": "61b75a820ddd6e00aff418b1",
    "email": "some@test.com",
    "isActivated": false,
  },
}
`;
